gpflow/utilities/utilities.py

```diff
@@ -131,45 +141,85 @@ def _merge_leaf_components(
+    inside the components of the tf.Module using `traverse_component`. Resets the caches stored inside each
+    tfp.bijectors.Bijector to allow the deepcopy of the tf.Module.
+
     :param module: tf.Module including keras.Model, keras.layers.Layer and gpflow.Module.
-    :param prefix: string containing the relative path to module, by default set to None.
     :return:
     """
-    if not isinstance(input, tf.Module):
-        raise TypeError("Input object expected to have `tf.Module` type")
-
-    prefix = input.__class__.__name__ if prefix is None else prefix
-    var_dict = dict()
-
-    for key, submodule in vars(input).items():
-        if key in tf.Module._TF_MODULE_IGNORED_PROPERTIES:
-            continue
-        elif isinstance(submodule, Parameter) or isinstance(submodule, tf.Variable):
-            var_dict[f"{prefix}.{key}"] = submodule
-        elif isinstance(submodule, tf.Module):
-            submodule_var = _get_leaf_components(submodule, prefix=f"{prefix}.{key}")
-            var_dict.update(submodule_var)
-        elif isinstance(submodule, ListWrapper):
-            submodule_name = input.__class__.__name__
-            for term_idx, subterm in enumerate(submodule):
-                subterm_key = f"{submodule_name}_{key}[{term_idx}]"
-                if isinstance(subterm, tf.Module):
-                    subterm_var = _get_leaf_components(subterm, prefix=f"{prefix}.{subterm_key}")
-                    var_dict.update(subterm_var)
-        elif isinstance(submodule, _DictWrapper):
-            submodule_name = input.__class__.__name__
-            for term_key, subterm in submodule.items():
-                subterm_key = f"{submodule_name}_{key}[{term_key}]"
-                subterm_var = _get_leaf_components(subterm, prefix=f"{prefix}.{subterm_key}")
-                var_dict.update(subterm_var)
-    return var_dict
+    target_types = (tfp.bijectors.Bijector,)
+    accumulator = ('', None)
+
+    def clear_bijector(bijector, _, state):
+        bijector._from_x.clear()
+        bijector._from_y.clear()
+        return state
+
+    _ = traverse_module(input, accumulator, clear_bijector, target_types)
+    return deepcopy(input)
+
+
+def traverse_module(m: TraverseInput,
+                    acc: Accumulator,
+                    update_cb: TraverseUpdateCallable,
+                    target_types: tuple) -> Accumulator:
+    """
+    Recursively traverses `m`, accumulating in `acc` a path and a state until it finds an object of type
+    in `target_types` to apply `update_cb` to update the accumulator `acc` and/or the object.
+
+    :param m: tf.Module, tf.Variable or gpflow.Parameter
+    :param acc: Tuple of path and state
+    :param update_cb: Callable
+    :param target_types: target class types
+    :return:
+    """
+    path, state = acc
+
+    new_state = state
+
+    if isinstance(m, target_types):
+        return update_cb(m, path, state)
+
+    if isinstance(m, ListWrapper):
```

Comment:
TL; DR - let's change `ListWrapper` and `_DictWrapper` to something thats not from the deepest guts of tensorflow. Open to discussion to what.

I'm not a huge fan of the usage of these non-exposed classes `ListWrapper` and `_DictWrapper` for a couple of reasons:
1. Since these classes are private (undocumented and buried), this is explicitly depending on implementation, not interface, leaving us with all the natural problems that arise from this. For one, comments on the implementation source, suggest this code is  [marked there for work](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/tracking/data_structures.py#L384)
2. I think we are effectively using these classes because the names "sound like" `list` and `dict` - not because we actually want to check instances of these specific classes. Can we be truer to our aim by either checking against `list` or `dict`, or (slightly worse) by checking for certain attributes  `has_attribute('items')`... The latter is slightly ugly (all dicts `has_attribute('items')`, not all `has_attribute('items')` are dicts) but it seems to me this is a true reflection of what actually want to check (do we _really_ care if the things is a dict?)

I ran the tests locally with `dict` and `list` and didnt see any different failures.

Reply:
I agree with this. I'm not a huge fan either. Perhaps, it can be addressed later on in a separate PR (?)

Reply:
I think given these are literally 2 word changes, that we should go for it with `list` and `dict`, unless @awav has a good reason why not. Here's why:
1. [ListWrapper](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/tracking/data_structures.py#L389) indicates that that it inherits from list, simply to allow `isinstance` checks!
2. [_DictWrapper](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/tracking/data_structures.py#L696) inherits from a `wrapt.ObjectProxy`, before going on to instantiate the proxy with a `dict`. The end result should still `isinstance(dict) == True` according to the [wrapt documentation](https://wrapt.readthedocs.io/en/latest/wrappers.html)

Reply:
```suggestion
    if isinstance(m, list):
```